consensus/ibft/ibft.go

```diff
@@ -452,6 +436,48 @@ func (i *Ibft) buildBlock(snap *Snapshot, parent *types.Header) (*types.Block, e
 	return block, nil
 }
 
+type transitionInterface interface {
+	Write(txn *types.Transaction) error
+}
+
+func (i *Ibft) writeTransactions(gasLimit uint64, transition transitionInterface) []*types.Transaction {
+	txns := []*types.Transaction{}
+	returnTxnFuncs := []func(){}
+	for {
+		txn, retTxnFn := i.txpool.Pop()
+		if txn == nil {
+			break
+		}
+
+		if txn.ExceedsBlockGasLimit(gasLimit) {
+			i.logger.Error(fmt.Sprintf("failed to write transaction: %v", state.ErrBlockLimitExceeded))
+			i.txpool.DecreaseAccountNonce(txn)
+			continue
+		}
+
+		if err := transition.Write(txn); err != nil {
+			if _, ok := err.(*state.GasLimitReachedTransitionApplicationError); ok {
+				returnTxnFuncs = append(returnTxnFuncs, retTxnFn)
+				break
+			} else if appErr, ok := err.(*state.TransitionApplicationError); ok && appErr.IsRecoverable {
+				returnTxnFuncs = append(returnTxnFuncs, retTxnFn)
+			} else {
+				i.txpool.DecreaseAccountNonce(txn)
+			}
+			continue
+		}
+
+		txns = append(txns, txn)
+	}
+
+	for _, retFunc := range returnTxnFuncs {
+		retFunc()
+	}
```

Comment:
I think it's clever that we run the callbacks later. Is there any side effect to having them run here, and not right after the txn fails?

I would love to have more in code documentation for this entire method 🙂

Reply:
We run them here and not immediately after the tnx fails because we don't want to return the txn to the pool and pop it again in the next iteration.

Reply:
[comments added](https://github.com/0xPolygon/polygon-sdk/pull/237/commits/ce819d1981811f3daae949691081ffdb827d48d8)